### PR TITLE
Better error messages for BatchNorm

### PR DESCRIPTION
**Pull request by fKunstner**
Makes the BatchNorm error message more explicit to avoid confusions (https://github.com/f-dangel/backpack/issues/239) and adds an option to ignore the exception.  

Summary of changes:

- Make BatchNorm error message [more explicit](https://github.com/f-dangel/backpack/commit/075834e64dff9f9297c75b4d5b51b9fcc3b5c405#diff-3b87a038266d15ca9769605a81763176d9670389ad809bea3771d43c8cc72c16R23-R28)
- Tie the BatchNorm raising of exception or warning [to fail_mode](https://github.com/f-dangel/backpack/commit/07dbbbfbab1d34b21a3d65aeb20a0da1c0bfefb9#diff-3b87a038266d15ca9769605a81763176d9670389ad809bea3771d43c8cc72c16R36-R41)
- Split error handling on missing extensions for first and second order
The current [error handling for missing modules](https://github.com/f-dangel/backpack/commit/73fd638c057c250a35e1cc1d6cd1a79325196a85#diff-cfc489564c02871bb06382d841a8a380544dce613ac05704deeb3cb31fc3a836L97-L111) mixed second-order extensions (which should fail if there is no extension defined) and first-order extensions (which should fail if there is no extension defined and the module has parameters). Moved to [first order](https://github.com/f-dangel/backpack/commit/73fd638c057c250a35e1cc1d6cd1a79325196a85#diff-9ad3a8ba51361c2bca2d17d5b7af5dd20320f1a23929f1563984775f0a7eaeebR33-R48) and [second order](https://github.com/f-dangel/backpack/commit/73fd638c057c250a35e1cc1d6cd1a79325196a85#diff-c05a11929412a47e0a2fe36222b048bae7a227e6e5b5625dd5f0ba102ecf4d75R13-R26) base classes.
- Change default fail_mode to error and make fail_mode user accessible
  First order extensions did not expose `fail_mode` and [had `warn` as a default](https://github.com/f-dangel/backpack/commit/c69dce1b69dca8fe5227bdc67065a9aa7db3f99a#diff-9ad3a8ba51361c2bca2d17d5b7af5dd20320f1a23929f1563984775f0a7eaeebR24).

### CODE REVIEW

**Comment by f-dangel** [comment]
Hey Fred, I skimmed through your changes:
1. The failing test checks if the result in `batch_grad` for a BN layer in train mode sums to `grad`. Working with `batch_grad` is 'okay' in this case because it's not interpreted as per-sample gradients. We could either revert the default for first-oder fail mode, or adapt the test to use `BatchGrad(fail_mode-"WARNING")`. I would currently favor to revert the default (as this also does not trigger a version bump, and fixes 2.).
2. The RTD example with the custom ResNet fails for similar reasons as in 1.
3. Can you `pip install --upgrade && make black` to update the formatting?

Happy to review or discuss!
